From 067986fa674f0811614dab4c4572f5f7ff483400 Mon Sep 17 00:00:00 2001
From: Nick Wellnhofer <wellnhofer@aevum.de>
Date: Sat, 18 Mar 2023 14:44:28 +0100
Subject: [PATCH] parser: Fix regressions from previous commits

- Fix memory leak in xmlParseNmtoken.
- Fix buffer overread after htmlParseCharDataInternal.
---
 HTMLparser.c | 2 +-
 parser.c     | 4 +++-
 2 files changed, 4 insertions(+), 2 deletions(-)

diff --git a/HTMLparser.c b/HTMLparser.c
index 0e79dffdd..ca3ebc413 100644
--- a/HTMLparser.c
+++ b/HTMLparser.c
@@ -3203,6 +3203,7 @@ htmlParseCharDataInternal(htmlParserCtxtPtr ctxt, int readahead) {
 	} else {
 	    COPY_BUF(l,buf,nbchar,cur);
 	}
+	NEXTL(l);
 	if (nbchar >= HTML_PARSER_BIG_BUFFER_SIZE) {
             buf[nbchar] = 0;
 
@@ -3228,7 +3229,6 @@ htmlParseCharDataInternal(htmlParserCtxtPtr ctxt, int readahead) {
 	    nbchar = 0;
             SHRINK;
 	}
-	NEXTL(l);
 	cur = CUR_CHAR(l);
     }
     if (ctxt->instate == XML_PARSER_EOF)
diff --git a/parser.c b/parser.c
index f69231d5a..8e548cdae 100644
--- a/parser.c
+++ b/parser.c
@@ -3683,8 +3683,10 @@ xmlParseNmtoken(xmlParserCtxtPtr ctxt) {
 		c = CUR_CHAR(l);
 	    }
 	    buffer[len] = 0;
-            if (ctxt->instate == XML_PARSER_EOF)
+            if (ctxt->instate == XML_PARSER_EOF) {
+                xmlFree(buffer);
                 return(NULL);
+            }
 	    return(buffer);
 	}
     }
